arangod/Cluster/AgencyCallback.cpp

```diff
@@ -64,25 +65,43 @@ void AgencyCallback::refetchAndUpdate(bool needToAcquireMutex, bool forceCheck)
     return;
   }
 
-  AgencyCommResult result = _agency.getValues(key);
-
-  if (!result.successful()) {
-    if (!_agency.server().isStopping()) {
-      // only log errors if we are not already shutting down...
-      // in case of shutdown this error is somewhat expected
-      LOG_TOPIC("fb402", ERR, arangodb::Logger::CLUSTER)
-          << "Callback getValues to agency was not successful: " << result.errorCode()
-          << " " << result.errorMessage();
+  VPackSlice result;
+  std::shared_ptr<VPackBuilder> builder;
+  consensus::index_t idx = 0;
+  if (local) {
+    auto& _cache = _agency.server().getFeature<ClusterFeature>().agencyCache();
+    std::tie(builder, idx) = _cache.get(std::vector<std::string>{AgencyCommManager::path(key)});
+    result = builder->slice();
+    if (!result.isArray()) {
+      if (!_agency.server().isStopping()) {
+        // only log errors if we are not already shutting down...
+        // in case of shutdown this error is somewhat expected
+        LOG_TOPIC("ec320", ERR, arangodb::Logger::CLUSTER)
+          << "Callback to get agency cache was not successful: " << result.toJson();
+      }
+      return;
     }
-    return;
+  } else {
+    AgencyCommResult tmp = _agency.getValues(key);
+    if (!tmp.successful()) {
+      if (!_agency.server().isStopping()) {
+        // only log errors if we are not already shutting down...
+        // in case of shutdown this error is somewhat expected
+        LOG_TOPIC("fb402", ERR, arangodb::Logger::CLUSTER)
+          << "Callback getValues to agency was not successful: " << tmp.errorCode()
+          << " " << tmp.errorMessage();
+      }
+      return;
+    }
+    result = tmp.slice();
```

Comment:
This looks dangerous, since `tmp` goes out of scope and `result` still points to it. I think this is broken.